src/Reserve.sol

```diff
@@ -37,44 +122,146 @@ contract Reserve is IReserve, Ownable {
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param plugin The plugin to withdraw from.
+    /// It doesn't need to be registered as a plugin for `token`.
+    /// Pass `NO_PLUGIN` to withdraw directly from the reserve balance.
+    /// @param to The address to which funds are withdrawn.
+    /// @param amt The withdrawn amount.
     function forceWithdraw(
         IERC20 token,
+        IReservePlugin plugin,
         address to,
         uint256 amt
     ) public onlyOwner {
-        require(token.transfer(to, amt), "Reserve: transfer failed");
-        emit ForceWithdrawn(msg.sender, token, to, amt);
+        _beforeWithdrawal(token, plugin, amt);
+        _transfer(token, _pluginAddr(plugin), to, amt);
+        emit ForceWithdrawn(msg.sender, token, plugin, to, amt);
+    }
+
+    /// @notice Sets the deposited amount counter for a token without transferring any funds.
+    /// If used recklessly, it may cause a mismatch between the counter and the actual balance
+    /// making valid future calls to `withdraw` or `setPlugin` fail due to lack of funds.
+    /// It may also make the counter lower than what users expect it to be again making
+    /// valid future calls to `withdraw` fail.
+    /// Callable only by the current owner.
+    /// @param token The used token.
+    /// @param amt The new deposited amount counter value.
+    function setDeposited(IERC20 token, uint256 amt) public onlyOwner {
+        deposited[token] = amt;
     }
 
+    /// @notice Adds a new user.
+    /// @param user The new user address.
     function addUser(address user) public onlyOwner {
         isUser[user] = true;
         emit UserAdded(msg.sender, user);
     }
 
+    /// @notice Removes an existing user.
+    /// @param user The removed user address.
     function removeUser(address user) public onlyOwner {
         isUser[user] = false;
         emit UserRemoved(msg.sender, user);
     }
+
+    function _afterDeposition(
+        IERC20 token,
+        IReservePlugin plugin,
+        uint256 amt
+    ) internal {
+        if (plugin != NO_PLUGIN) plugin.afterDeposition(token, amt);
+    }
+
+    function _beforeWithdrawal(
+        IERC20 token,
+        IReservePlugin plugin,
+        uint256 amt
+    ) internal {
+        if (plugin != NO_PLUGIN) plugin.beforeWithdrawal(token, amt);
+    }
+
+    function _pluginAddr(IReservePlugin plugin) internal view returns (address) {
+        return plugin == NO_PLUGIN ? address(this) : address(plugin);
+    }
+
+    function _transfer(
+        IERC20 token,
+        address from,
+        address to,
+        uint256 amt
+    ) internal {
+        bool success;
+        if (from == address(this)) {
+            success = token.transfer(to, amt);
+        } else {
+            success = token.transferFrom(from, to, amt);
+        }
+        require(success, "Reserve: transfer failed");
+    }
```

Comment:
We have already the global view function for `deposited(IERC20 token)` it would be nice to add one for the `balance` as well.

Basically a wrapper function which asks the balance of the plugin. In case of no plugin it might be `erc20.blanceOf(this)`.

Reply:
I think that `balance()` is a good idea if we add the surplus support to `Reserve`. Otherwise it's not very useful, because it doesn't facilitate the usage of the `Reserve` API. It also adds a function to the plugin API which by its design makes an assumption about the balance is a single, readily available number. It also implies that this exact amount to be immediately withdrawable, which may not be true, because funds could be available only in the future.

Reply:
> It also adds a function to the plugin API which by its design makes an assumption about the balance being a single, readily available number.

Yes, but i think this a requirement. If some Plugin can not offer that, it might be not possible to withdraw at any point in time. Which we definitely need as a functionality.

I think it might be later useful to query certain information centralized.

It was just an idea.